client/comms/wsconn.go

```diff
@@ -0,0 +1,349 @@
+type WsConn struct {
+	rID          uint64
+	cfg          *WsCfg
+	ws           *websocket.Conn
+	tlsCfg       *tls.Config
+	readCh       chan *msgjson.Message
+	sendCh       chan *msgjson.Message
+	req          map[uint64]*msgjson.Message
+	reqMtx       sync.RWMutex
+	writeMtx     sync.Mutex
+	connected    bool
+	connectedMtx sync.RWMutex
+	once         sync.Once
+	wg           sync.WaitGroup
+}
+
+// filesExists reports whether the named file or directory exists.
+func fileExists(name string) bool {
+	_, err := os.Stat(name)
+	return !os.IsNotExist(err)
+}
+
+// NewWsConn creates a client websocket connection.
+func NewWsConn(cfg *WsCfg) (*WsConn, error) {
+	if cfg.PingWait < 1 {
+		return nil, fmt.Errorf("ping wait time should be a second or greater")
+	}
+
+	if cfg.WriteWait < 1 {
+		return nil, fmt.Errorf("write wait time should be a second or greater")
+	}
+
+	var tlsConfig *tls.Config
+	if fileExists(cfg.RpcCert) && fileExists(cfg.RpcKey) {
+		keypair, err := tls.LoadX509KeyPair(cfg.RpcCert, cfg.RpcKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// Prepare the TLS configuration.
+		tlsConfig = &tls.Config{
+			Certificates: []tls.Certificate{keypair},
+			MinVersion:   tls.VersionTLS12,
+		}
+
+		if cfg.InsecureSkipVerify {
+			tlsConfig.InsecureSkipVerify = true
+		}
+	}
+
+	return &WsConn{
+		cfg:    cfg,
+		tlsCfg: tlsConfig,
+		readCh: make(chan *msgjson.Message, readBuffSize),
+		sendCh: make(chan *msgjson.Message),
+		req:    make(map[uint64]*msgjson.Message),
+	}, nil
+}
+
+// isConnected returns the connection connected state.
+func (conn *WsConn) isConnected() bool {
+	conn.connectedMtx.RLock()
+	defer conn.connectedMtx.RUnlock()
+	return conn.connected
+}
+
+// setConnected updates the connection's connected state
+func (conn *WsConn) setConnected(connected bool) {
+	conn.connectedMtx.Lock()
+	conn.connected = connected
+	conn.connectedMtx.Unlock()
+}
+
+// logRoute logs a request keyed by its id.
+func (conn *WsConn) logRequest(id uint64, req *msgjson.Message) {
+	conn.reqMtx.Lock()
+	conn.req[id] = req
+	conn.reqMtx.Unlock()
+}
+
+// FetchRequest fetches the request associated with the id. The returned
+// request is removed from the cache.
+func (conn *WsConn) FetchRequest(id uint64) (*msgjson.Message, error) {
+	conn.reqMtx.Lock()
+	defer conn.reqMtx.Unlock()
+	req := conn.req[id]
+	if req == nil {
+		return nil, fmt.Errorf("no request found for id %d", id)
+	}
+
+	delete(conn.req, id)
+
+	return req, nil
+}
+
+// NextID returns the next request id.
+func (conn *WsConn) NextID() uint64 {
+	return atomic.AddUint64(&conn.rID, 1)
+}
+
+// connect attempts to establish a websocket connection.
+func (conn *WsConn) connect() error {
+	scheme := "ws"
+
+	if conn.tlsCfg != nil {
+		scheme = "wss"
+	}
+
+	url := url.URL{
+		Scheme: scheme,
+		Host:   conn.cfg.Host,
+		Path:   conn.cfg.Path,
+	}
+
+	dialer := &websocket.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: 10 * time.Second,
+	}
```

Comment:
That'll work for now.  A proxy config will almost certainly get added later, but let's not get stuck on that.